internal/sync2/keylock.go

```diff
@@ -0,0 +1,65 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information
+
+package sync2
+
+import (
+	"sync"
+)
+
+// KeyLock provides per-key RW locking. Locking is key-specific, meaning lock
+// contention only exists for locking calls using the same key parameter. As
+// with all locks, do not call Unlock() or RUnlock() before a corresponding
+// Lock() or RLock() call.
+//
+// Note on memory usage: internally KeyLock lazily and atomically creates a
+// separate sync.RWMutex for each key. To maintain synchronization guarantees,
+// these interal mutexes are not freed until the entire KeyLock instance is
+// freed.
+type KeyLock struct {
+	locks sync.Map
+}
+
+// NewKeyLock create a new KeyLock
+func NewKeyLock() *KeyLock {
+	return &KeyLock{}
+}
+
+// Lock the provided key
+func (l *KeyLock) Lock(key interface{}) {
+	lock := l.getNewLock(key)
+	lock.Lock()
+}
+
+// Unlock the provided key
+func (l *KeyLock) Unlock(key interface{}) {
+	lock := l.getExistingLock(key)
+	lock.Unlock()
+}
+
+// RLock the provided key
+func (l *KeyLock) RLock(key interface{}) {
+	lock := l.getNewLock(key)
+	lock.RLock()
+}
+
+// RUnlock the provided key
+func (l *KeyLock) RUnlock(key interface{}) {
```

Comment:
When you return the unlocking func from the locking then you can avoid the double lookup.

Reply:
Done.